DigitalLearningSolutions.Data/Services/UserService.cs

```diff
@@ -242,9 +259,77 @@ public IEnumerable<DelegateUserCard> GetDelegateUserCardsForWelcomeEmail(int cen
             );
         }
 
+        public void UpdateAdminUserPermissions(
+            int adminId,
+            bool isCentreAdmin,
+            bool isSupervisor,
+            bool isTrainer,
+            bool isContentCreator,
+            bool isContentManager,
+            bool importOnly,
+            int categoryId
+        )
+        {
+            if (NewUserRolesExceedAvailableSpots(adminId, isTrainer, isContentCreator, isContentManager, importOnly))
+            {
+                throw new AdminRoleFullException(
+                    "Failed to update admin roles for admin " + adminId +
+                    " as one or more of the roles being added to have reached their limit"
+                );
+            }
+
+            userDataService.UpdateAdminUserPermissions(
+                adminId,
+                isCentreAdmin,
+                isSupervisor,
+                isTrainer,
+                isContentCreator,
+                isContentManager,
+                importOnly,
+                categoryId
+            );
+        }
+
         private static bool UserEmailHasChanged(User? user, string emailAddress)
         {
             return user != null && user.EmailAddress != emailAddress;
         }
+
+        private bool NewUserRolesExceedAvailableSpots(
+            int adminId,
+            bool isTrainer,
+            bool isContentCreator,
+            bool isContentManager,
+            bool importOnly
+        )
+        {
+            var oldUserDetails = userDataService.GetAdminUserById(adminId)!;
+            var currentNumberOfAdmins =
+                centreContractAdminUsageService.GetCentreAdministratorNumbers(oldUserDetails.CentreId);
+
+            if (isTrainer && !oldUserDetails.IsTrainer && currentNumberOfAdmins.TrainersAtOrOverLimit)
+            {
+                return true;
+            }
+
+            if (isContentCreator && !oldUserDetails.IsContentCreator && currentNumberOfAdmins.CcLicencesAtOrOverLimit)
+            {
+                return true;
+            }
+
+            if (isContentManager && importOnly & !oldUserDetails.IsCmsAdministrator &&

```

Comment:
If we wrap this in AdminRoles model, we can have a derived field cmsAdministrator on it (which will be nicer to read here over `isContentManager && importOnly` )